src/gpu/d3d11/SDL_gpu_d3d11.c

```diff
@@ -3733,28 +3746,432 @@ static SDL_GpuCommandBuffer* D3D11_AcquireCommandBuffer(
+                        uniformBufferAlreadyBound = SDL_TRUE;
+                        break;
+                    }
+                }
+
+                /* If not already bound by this command buffer, cycle and reset offset */
+                /* This is because the first map call on a deferred context must be DISCARD */
+                if (!uniformBufferAlreadyBound)
+                {
+                    D3D11_INTERNAL_CycleActiveBuffer(
+                        renderer,
+                        uniformBuffer->bufferContainer
+                    );
+
+                    D3D11_INTERNAL_TrackBuffer(
+                        d3d11CommandBuffer,
+                        uniformBuffer->bufferContainer->activeBuffer
+                    );
+
+                    uniformBuffer->offset = 0;
+                }
+                break;
+
+            default:
+                SDL_LogError(SDL_LOG_CATEGORY_APPLICATION, "Unsupported shader resource type!");
+                break;
+        }
+    }
+}
+
+static void D3D11_INTERNAL_PushUniformData(
+    SDL_GpuCommandBuffer *commandBuffer,
+    SDL_GpuUniformBuffer *uniformBuffer,
+    void *data,
+    Uint32 dataLengthInBytes
+) {
+    D3D11CommandBuffer *d3d11CommandBuffer = (D3D11CommandBuffer*) commandBuffer;
+    D3D11Renderer *renderer = d3d11CommandBuffer->renderer;
+    D3D11UniformBuffer *d3d11UniformBuffer = (D3D11UniformBuffer*) uniformBuffer;
+    ID3D11Buffer *nullBuf = NULL;
+    Uint32 offsetInConstants, blockSizeInConstants;
+    Uint32 drawOffset;
+    Uint32 i;
+
+    if (d3d11UniformBuffer->offset + d3d11UniformBuffer->currentBlockSize >= d3d11UniformBuffer->bufferContainer->activeBuffer->size)
+    {
+        D3D11_INTERNAL_CycleActiveBuffer(
+            renderer,
+            d3d11UniformBuffer->bufferContainer
+        );
+
+        d3d11UniformBuffer->offset = 0;
+
+        D3D11_INTERNAL_TrackBuffer(
+            d3d11CommandBuffer,
+            d3d11UniformBuffer->bufferContainer->activeBuffer
+        );
+    }
+
+    drawOffset = d3d11UniformBuffer->offset;
+
+    D3D11_INTERNAL_SetUniformBufferData(
+        renderer,
+        d3d11CommandBuffer,
+        d3d11UniformBuffer->bufferContainer->activeBuffer,
+        d3d11UniformBuffer->offset,
+        data,
+        dataLengthInBytes
+    );
+
+    d3d11UniformBuffer->offset += d3d11UniformBuffer->currentBlockSize;
 
-	commandBuffer = D3D11_INTERNAL_GetInactiveCommandBufferFromPool(renderer);
-	commandBuffer->windowData = NULL;
-	commandBuffer->graphicsPipeline = NULL;
-	commandBuffer->computePipeline = NULL;
-	commandBuffer->vertexUniformBuffer = NULL;
-	commandBuffer->fragmentUniformBuffer = NULL;
-	commandBuffer->computeUniformBuffer = NULL;
-	for (Uint32 i = 0; i < MAX_COLOR_TARGET_BINDINGS; i += 1)
-	{
-		commandBuffer->colorTargetResolveTexture[i] = NULL;
-		commandBuffer->colorTargetResolveSubresourceIndex[i] = 0;
-		commandBuffer->colorTargetMsaaHandle[i] = NULL;
-	}
+    offsetInConstants = drawOffset / 16;
```

Comment:
`(drawOffset % 16) == 0` is something i would assert on here just to be able to sleep at night

Reply:
drawOffset is always aligned to the block size which is aligned to 256